kia-lib/src/main/kotlin/com/faforever/ice/telemetry/TelemetryClient.kt

```diff
@@ -51,35 +61,66 @@ class TelemetryClient(
     init {
         logger.info {
             "Open the telemetry ui via ${
-                serverBaseUrl.replaceFirst(
-                    "ws",
-                    "http",
-                )
+                serverBaseUrl.replaceFirst("ws", "http")
             }/app.html?gameId=${iceOptions.gameId}&playerId=${iceOptions.userId}"
         }
 
         val uri: URI = URI.create("$serverBaseUrl/adapter/v1/game/${iceOptions.gameId}/player/${iceOptions.userId}")
         websocketClient = TelemetryWebsocketClient(uri)
 
-        connectingFuture = connectAsync()
+        sendingLoopThread = Thread(this::sendingLoop)
+        sendingLoopThread.uncaughtExceptionHandler = Thread.UncaughtExceptionHandler { _: Thread?, e: Throwable? ->
+            logger.error { "Thread sendingLoop crashed unexpectedly: $e" }
+        }
+
+        CompletableFuture.runAsync {
+            websocketClient.connectBlocking()
+            sendingLoopThread.start()
+        }
+
         registerAsPeer()
     }
 
-    private fun connectAsync() = CompletableFuture.runAsync(websocketClient::connect)
-        .exceptionally {
-            logger.error(it) { "Failed to connect to telemetry websocket" }
-            null
+    private fun sendingLoop() {
+        while (!connectionFailed) {
+            val message = messageQueue.take()
+
+            if (websocketClient.isClosed) {
+                Failsafe.with(
+                    RetryPolicy.builder<ConnectionResult>()
+                        .handleResult(ConnectionResult.FAILURE)
+                        .withBackoff(Duration.ofSeconds(2), Duration.ofMinutes(1))
+                        .withMaxRetries(connectionRetries)
```

Comment:
Current logic is 6 retries with exponential backoff starting at 2 seconds and going up to 1 minute. At 6 retries this currently means we give up after a couple minutes. Just my best guess, let me know if you have something else in mind.

Reply:
That should be fine.